### ISSUE TITLE
`twistd mail --auth` ineffective

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5219]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5219|
|Type|release blocker: regression|
|Created|2011-08-05 18:10:50Z|
|Branch|https://github.com/twisted/twisted/tree/mail-auth-5219|

[#4740](https://github.com/twisted/twisted/issues/4740) added `--auth` (and the related options) command line option to `twistd mail` and deprecated the old explicit password-file based usage.  However, it didn't add any code to use the checkers created by `--auth` to configure a portal.  Thus, the options are useless in their current form.

The checkers put into the `'credCheckers'` key of the options object need to be attached a portal object, similar to what the deprecated `opt_passwordfile` does with the checker it creates.

Attachments:

* [twisted5219.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afc/afc22e9ff9923f322d32bb744d8e392c39d8378c/7982301a71f1e300de70a120125caf9561e516a0.patch) (540 bytes) - added by tpratt on 2011-08-05 19:05:30Z - 
* [twisted5219_with_test.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afc/afc22e9ff9923f322d32bb744d8e392c39d8378c/a8ba3980d1aae8b0aa43cd48bc8f6cfb33e70cdf.patch) (1509 bytes) - added by necaris on 2012-03-12 21:14:44Z - Patch with unit test
* [twisted5219_with_test.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/afc/afc22e9ff9923f322d32bb744d8e392c39d8378c/9a40cdf177f6b9e6240b6ce659580c40d4d7b923.patch) (1698 bytes) - added by necaris on 2012-03-13 03:34:34Z - Updated patch with test

<details><summary>Searchable metadata</summary>

```
trac-id__5219 5219
type__release_blocker__regression release blocker: regression
reporter__exarkun exarkun
priority__high high
milestone__ 
branch__branches_mail_auth_5219 branches/mail-auth-5219
branch_author__cyli cyli
status__closed closed
resolution__fixed fixed
component__mail mail
keywords__ 
time__1312567850000000 1312567850000000
changetime__1331633861000000 1331633861000000
version__None None
owner__ 
cc__necaris
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun set owner to tpratt|
|-|-|

Thanks.  Can you add some unit tests too?

**Comment by twisted-trac** [comment]
|<img alt="necaris's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|necaris removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|<img alt="arsenerei's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|arsenerei commented|
|-|-|

Unfortunately, a number of tests fail. I wrapped the self['credCheckers'] logic in try/except blocks so that all the failing tests would pass, but the test_auth test still failed. I don't know enough of this code to know how to fix that.

**Comment by twisted-trac** [comment]
|<img alt="necaris's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|necaris commented|
|-|-|

arsenerei: thanks for the review -- I'd been assuming the original patch fixed the issue, and that all that was needed was a test. It turns out, of course, that the original patch needed a slight modification -- this and the updated test are attached in the new patch.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

(In [33725]) Branching to 'mail-auth-5219'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [33726]) Apply patch from necaris

refs [#5219](https://github.com/twisted/twisted/issues/5219)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli commented|
|-|-|

Thanks so much for working on this necaris!  

This looks great ([buildbot reuslts](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/mail-auth-5219))!  The one thing I'd keep in mind for future patches is that you should write a [news file](http://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) for every ticket.  Please see the link for more details about the format and the name of the file, but in short it's file that contains a description of the change made in the ticket. For every release of Twisted, the descriptions in these files will be listed into the release notes.

For this ticket I'll make one, and I'll merge your patch in.

Once again, thanks!

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_9"></a>|@cyli commented|
|-|-|

(In [33728]) Add news file

refs [#5219](https://github.com/twisted/twisted/issues/5219)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_10"></a>|@cyli set status to `closed`|
|-|-|

(In [33729]) Merge mail-auth-5219: Fix twistd mail --auth

Author: necaris
Reviewer: arsenerei, cyli
Fixes: [#5219](https://github.com/twisted/twisted/issues/5219)

Attaches checkers created by the options object to the portal object in twisted.mail.tap.py

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_11"></a>|@cyli commented|
|-|-|

Note - since it has deprecated since 11.0, and has been broken anyway, I have filed [#5541](https://github.com/twisted/twisted/issues/5541) to remove the --paswordfile option.
